diff --git a/modules/taxonomy/views_handler_filter_term_node_tid.inc b/modules/taxonomy/views_handler_filter_term_node_tid.inc
index 5ca6d18..557ec39 100644
--- a/modules/taxonomy/views_handler_filter_term_node_tid.inc
+++ b/modules/taxonomy/views_handler_filter_term_node_tid.inc
@@ -195,7 +195,6 @@ class views_handler_filter_term_node_tid extends views_handler_filter_many_to_on
         '#options' => $options,
         '#size' => min(9, count($options)),
         '#default_value' => $default_value,
-        '#description' => t('Leave blank for all. Otherwise, the first selected term will be the default instead of "Any".'),
       );
 
       if (!empty($form_state['exposed']) && isset($identifier) && !isset($form_state['input'][$identifier])) {
@@ -207,6 +206,8 @@ class views_handler_filter_term_node_tid extends views_handler_filter_many_to_on
     if (empty($form_state['exposed'])) {
       // Retain the helper option
       $this->helper->options_form($form, $form_state);
+
+      $form['value']['#description'] = t('Leave blank for all. Otherwise, the first selected term will be the default instead of "Any".');
     }
   }
 
