diff --git a/geocoder.widget.inc b/geocoder.widget.inc
index 9e350a6..b4cf41b 100644
--- a/geocoder.widget.inc
+++ b/geocoder.widget.inc
@@ -297,23 +297,31 @@ function geocoder_field_attach_presave($entity_type, $entity) {
   // Loop over any geofield using our geocode widget
   foreach ($instances as $field_name => $instance) {
     if ($instance['widget']['type'] === 'geocoder'
-      && !empty($entity->{$field_name})
       && ($geocoded_value = geocoder_widget_get_field_value($entity_type, $instance, $entity))) {
-      foreach ($entity->{$field_name} as $langcode => $items) {
-        if (empty($items)) {
-          $entity->{$field_name} = $geocoded_value;
-          break;
-        }
-        foreach ($items as $delta => $item) {
-          // Check the entity value.
-          if (!(isset($entity->geocoder_overridden)
-            && isset($entity->geocoder_overridden[$field_name])
-            && isset($entity->geocoder_overridden[$field_name][$langcode])
-            && !empty($entity->geocoder_overridden[$field_name][$langcode][$delta]))) {
-            $entity->{$field_name}[$langcode][$delta] = $geocoded_value[$langcode][$delta];
+
+      // If geofield is empty, it probably means that we chose
+      // 'Geocode from another field' widget. So we put geocoded value into
+      // this field.
+      if (!empty($entity->{$field_name})) {
+        foreach ($entity->{$field_name} as $langcode => $items) {
+          if (empty($items)) {
+            $entity->{$field_name} = $geocoded_value;
+            break;
+          }
+          foreach ($items as $delta => $item) {
+            // Check the entity value.
+            if (!(isset($entity->geocoder_overridden)
+              && isset($entity->geocoder_overridden[$field_name])
+              && isset($entity->geocoder_overridden[$field_name][$langcode])
+              && !empty($entity->geocoder_overridden[$field_name][$langcode][$delta]))) {
+              $entity->{$field_name}[$langcode][$delta] = $geocoded_value[$langcode][$delta];
+            }
           }
         }
       }
+      else {
+        $entity->{$field_name} = $geocoded_value;
+      }
     }
   }
 }
@@ -355,7 +363,8 @@ function geocoder_widget_get_entity_field_value($entity_type, $field_instance, $
   else if ($entity) {
     $wrapper = entity_metadata_wrapper($entity_type, $entity);
     $field_wrapper = $wrapper->value()->$field_name;
-    $value = $field_wrapper[LANGUAGE_NONE];
+    $langcode = field_language($entity_type, $entity, $field_name);
+    $value = $field_wrapper[$langcode];
     $values = array_filter(is_array($value) && isset($value[0]) ? $value : array($value));
     $source_field_values = array_map(function($value) {
       if (is_array($value)) {
